### PR TITLE
Interpreter FFI: custom compile_dir, use /tmp/ by default, and delete after execution

### LINKED ISSUE #1974: nit interpreter let some compile dir in the directory
Some `nit_compile` directory is not cleaned

~~
$ du -sh nit_compile
du: cannot access 'nit_compile': No such file or directory
$ nit examples/hello_world.nit
hello world
$ du -sh nit_compile
380K	nit_compile
~~~

### PR DESCRIPTION
**Pull request by xymus**
Generate native files in the `/tmp/` directory and delete them after execution. The directory can be customized using `--compile-dir`.

This should fix #1974 and #1975.

To support some other OS, we will need to use something other than `getpid` and `/tmp/`. We could also improve upon this by reusing the generated libraries between executions.

### CODE REVIEW

File: src/interpreter/dynamic_loading_ffi/on_demand_compiler.nit [hunk 1]
```diff
@@ -45,10 +54,24 @@ redef class AMethPropdef
...
+	end
+
 	# Where to store generated C and extracted code
-	#
-	# TODO make customizable and delete when execution completes
-	private var compile_dir = "nit_compile"
+	private var compile_dir: String is lazy do
+		var opt = modelbuilder.toolcontext.opt_compile_dir.value
+		if opt != null then return opt
+		return "/tmp/niti_ffi_{process_id}"
```
Comment (privat) [inline src/interpreter/dynamic_loading_ffi/on_demand_compiler.nit]: I do not like the possible `leaking` information when storing things in a world readable and writable place. Maybe `$XDG_RUNTIME_DIR` ? It is safer than /tmp I think.

**Comment by xymus** [comment]
Added a commit to prioritize using XDG_RUNTIME_DIR before /tmp/.

**Comment by privat** [comment]
+1

**Comment by privat** [comment]
test failed

**Comment by xymus** [comment]
On the server, `/run/user` is `noexec`, so shared libraries cannot be loaded. Any idea how to detect if a folder can hold executable code?

**Comment by privat** [comment]
Then /run may be a bad place, sorry for the suggestion. The spec do not
talk about executability but expect an access mode of 700
See
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

It is also possible that the environment variable is poorly set on the test
machine and thus fail. Per specification, an non absolute path should be
ignored.

I propose that you stay with /tmp for the moment. Just try to create the
directory with 0700. since we now have optional parameters, you can extend
mkdir with a mode :)

Le dim. 6 mars 2016 08:00, Alexis Laferrière <notifications@github.com> a
écrit :

**Comment by xymus** [comment]
Added the optional argument to mkdir and removed the commit using `/run/user/`.

**Pull request merged**